decompiler/pipeline/controlflowanalysis/expression_simplification_rules.py

```diff
@@ -0,0 +1,145 @@
+from decompiler.structures.visitors.substitute_visitor import SubstituteVisitor
+from decompiler.task import DecompilerTask
+
+
+class _ExpressionSimplificationRulesBase(PipelineStage, ABC):
+
+    def run(self, task: DecompilerTask):
+        max_iterations = task.options.getint("expression-simplification.max_iterations")
+        simplify_instructions(self._get_instructions(task), max_iterations)
+
+    @abstractmethod
+    def _get_instructions(self, task: DecompilerTask) -> list[Instruction]:
+        pass
+
+
+class ExpressionSimplificationRulesCfg(_ExpressionSimplificationRulesBase):
+    """
+    Pipeline stage that simplifies cfg expressions by applying a set of simplification rules.
+    """
+
+    name = "expression-simplification-rules-cfg"
+
+    def _get_instructions(self, task: DecompilerTask) -> list[Instruction]:
+        return list(task.graph.instructions)
+
+
+class ExpressionSimplificationRulesAst(_ExpressionSimplificationRulesBase):
+    """
+    Pipeline stage that simplifies ast expressions by applying a set of simplification rules.
+    """
+
+    name = "expression-simplification-rules-ast"
+
+    def _get_instructions(self, task: DecompilerTask) -> list[Instruction]:
+        instructions = []
+        for node in task.syntax_tree.topological_order():
+            if isinstance(node, CodeNode):
+                instructions.extend(node.instructions)
+
+        return instructions
+
+
+_pre_rules: list[SimplificationRule] = []
+_rules: list[SimplificationRule] = [
+    TermOrder(),
+    SubToAdd(),
+    SimplifyRedundantReference(),
+    SimplifyTrivialArithmetic(),
+    SimplifyTrivialBitArithmetic(),
+    SimplifyTrivialLogicArithmetic(),
+    SimplifyTrivialShift(),
+    CollapseConstants(),
+    CollectTerms(),
+]
+_post_rules: list[SimplificationRule] = [
+    CollapseAddNeg(),
+    FixAddSubSign()
+]
+
+
+def simplify_instructions(instructions: list[Instruction], max_iterations: int):
+    rule_sets = [
+        ("pre-rules", _pre_rules),
+        ("rules", _rules),
+        ("post-rules", _post_rules)
+    ]
+    for rule_name, rule_set in rule_sets:
+        iteration_count = _simplify_instructions_with_rule_set(instructions, rule_set, max_iterations)
+        if iteration_count <= max_iterations:
+            logging.info(f"Expression simplification took {iteration_count} iterations for {rule_name}")
+        else:
+            logging.warning(f"Exceeded max iteration count for {rule_name}")
+
+
+def _simplify_instructions_with_rule_set(
+        instructions: list[Instruction],
+        rule_set: list[SimplificationRule],
+        max_iterations: int
+) -> int:
+    iteration_count = 0
+
+    changes = True
+    while changes:
+        changes = False
+
+        for rule in rule_set:
+            for instruction in instructions:
+                for expression in instruction.subexpressions():
+                    while True:
+                        if expression is None:
```

Comment:
I initial didn't extract this to a function, because the inner looped needed read/write access to `changes` and `iteration_count`. So extracting that code needed some additional code for handling these variables. Fixed with e8992dcc.